pkg/common/undo.go

```diff
@@ -0,0 +1,43 @@
+package common
+
+// UndoFn is a function with no arguments which returns an error or nil.
+type UndoFn func() error
+
+// UndoStack models a simple undo stack which consumers can use to store undo
+// stateful functions, such as a function to teardown API state if something
+// goes wrong during procedural commands, for example deleting a Fastly service
+// after it's been created.
+type UndoStack struct {
+	states []UndoFn
+}
+
+// NewUndoStack constructs a new UndoStack.
+func NewUndoStack() *UndoStack {
+	s := make([]UndoFn, 0, 1)
+	stack := &UndoStack{
+		states: s,
+	}
+	return stack
+}
+
+// Pop method pops last added UndoFn element oof the stack and returns it.
+// If stack is empty Pop() returns nil.
+func (s *UndoStack) Pop() UndoFn {
+	n := len(s.states)
+	if n == 0 {
+		return nil
+	}
+	v := s.states[n-1]
+	s.states = s.states[:n-1]
+	return v
+}
+
+// Push method pushes an Undoer element onto the UndoStack.
+func (s *UndoStack) Push(elem UndoFn) {
+	s.states = append(s.states, elem)
+}
+
+// Len method returns the number of elements in the UndoStack.
+func (s *UndoStack) Len() int {
+	return len(s.states)
+}
```

Comment:
If you are going to go through the (appreciated!) trouble of defining an UndoStack type, IMO you should go all the way and also provide a helper method to execute it in case of error.

```go
func (s *UndoStack) RunIfError(err error) {
	if err == nil {
		return
	}
	for i := len(s.states)-1; i >= 0; i-- {
		s.states[i]()
	}
}
```

That means you could reduce the public API to merely `Push` and `RunIfError`, probably remove the UndoFn type and use plain `func()`, and use it in-situ like

```go
func f() (err error) {
	var undo common.UndoStack
	defer undo.RunIfError(err)
	
	undo.Push(func() { ... })
```

Reply:
If the undo funcs returned an error, RunIfError could take an io.Writer to which error results could be printed.

Reply:
Added this helper method in https://github.com/fastly/cli/pull/5/commits/d4b73e21737f74615747e148b2902b1ccaca7ff2